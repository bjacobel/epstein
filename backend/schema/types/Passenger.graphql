scalar AWSURL

type Passenger implements Node {
  id: Int  # 'literal' passengers will not have ids. ids are only given when they are canonicalized
  literal: String!
  name: String
  biography: String
  wikipedia_link: AWSURL
  image: AWSURL
  flights: FlightConnection
}
