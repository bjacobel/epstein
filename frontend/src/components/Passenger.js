import React from 'react';
import { useQuery, gql } from '@apollo/client';

const PASSENGER = gql`
  query Passenger($slug: String!) {
    passenger(slug: $slug) {
      id
      name
      biography
      wikipedia_link
      image
      flights(limit: 10) {
        edges {
          id
        }
      }
    }
  }
`;

export default ({ match }) => {
  const { slug } = match.params;
  const { loading, error, data } = useQuery(PASSENGER, { variables: { slug } });

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  return <pre>{JSON.stringify(data, null, 2)}</pre>;
};
