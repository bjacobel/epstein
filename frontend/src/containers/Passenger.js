import React from 'react';
import { useQuery, gql } from '@apollo/client';

import Details from '../components/Details';
import PaginatedBrowser from '../components/PaginatedBrowser';
import Loading from '../components/Loading';
import MiniFlight from '../components/Mini/MiniFlight';
import { link } from '../stylesheets/shared.css';

const FLIGHT_LIMIT = 10;

const PASSENGER = gql`
  query Passenger($slug: String!, $offset: Int, $limit: Int!) {
    passenger(slug: $slug) {
      id
      name
      biography
      wikipedia_link
      image
      flightCount
      flights(offset: $offset, limit: $limit) {
        edges {
          id
        }
        pageInfo {
          count
          hasNext
        }
      }
    }
  }
`;

const range = histogram => {
  if (!histogram) return '';

  const months = Object.keys(histogram);
  const [firstDate] = months[0];
  const [lastDate] = months[months.length - 1];

  if (firstDate === lastDate) {
    return `in ${firstDate}`;
  } else {
    return `between ${firstDate} and ${lastDate}`;
  }
};

export default ({ match }) => {
  const { slug } = match.params;
  const { loading, error, data, fetchMore } = useQuery(PASSENGER, {
    variables: { slug, limit: FLIGHT_LIMIT },
  });

  if (loading) return <Loading text />;
  if (error) return <p>Error :(</p>;

  return (
    <Details>
      <span>name</span>
      <span>{data.passenger.name}</span>
      <span>bio</span>
      <div>
        <span>{data.passenger.biography}</span>
        <br />
        <a className={link} href={data.passenger.wikipedia_link}>
          <span>See more on Wikipedia â†’</span>
        </a>
      </div>
      <span>history</span>
      <div>
        <span>{`Appears in the flight logs at least ${data.passenger.flightCount} times`}</span>
        <span>{range(data.passenger.histogram)}</span>
      </div>
      {data.passenger.image ? (
        <>
          <span>image</span>
          <img src={data.passenger.image} alt="" />
        </>
      ) : null}
      <span>flights</span>
      <PaginatedBrowser
        browserComponent={MiniFlight}
        ids={data.passenger.flights.edges.map(x => x.id)}
        totalAvailable={data.passenger.flightCount}
        pageSize={FLIGHT_LIMIT}
        fetchMore={() =>
          fetchMore({
            variables: { offset: data.passenger.flights.edges.length + 1 },
            updateQuery: (prev, { fetchMoreResult }) => {
              if (!fetchMoreResult) return prev;
              return {
                passenger: {
                  ...prev.passenger,
                  flights: {
                    ...fetchMoreResult.passenger.flights,
                    pageInfo: {
                      ...fetchMoreResult.passenger.flights.pageInfo,
                      count:
                        prev.passenger.flights.pageInfo.count +
                        fetchMoreResult.passenger.flights.pageInfo.count,
                    },
                    edges: [
                      ...prev.passenger.flights.edges,
                      ...fetchMoreResult.passenger.flights.edges,
                    ],
                  },
                },
              };
            },
          })
        } // eslint-disable-line
      />
    </Details>
  );
};
