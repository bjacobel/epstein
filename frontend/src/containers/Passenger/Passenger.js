import React from 'react';
import { useQuery, gql } from '@apollo/client';
import { format, parseISO, isEqual, isSameYear } from 'date-fns';

import Details from '../../components/Details';
import PaginatedBrowser from '../../components/PaginatedBrowser';
import Loading from '../../components/Loading';
import MiniFlight from '../../components/Mini/MiniFlight';
import { link } from '../../stylesheets/shared.css';
import MetaTags from '../../components/MetaTags';

const FLIGHT_LIMIT = 10;

export const PASSENGER = gql`
  query Passenger($slug: String!, $offset: Int, $limit: Int!) {
    passenger(slug: $slug) {
      id
      name
      slug
      biography
      wikipedia_link
      image
      flightCount
      histogram {
        count
        month
      }
      flights(offset: $offset, limit: $limit) {
        edges {
          id
        }
        pageInfo {
          count
          hasNext
        }
      }
    }
  }
`;

export const range = histogram => {
  if (!histogram.length) return '';
  const firstDate = parseISO(histogram[0].month);
  const lastDate = parseISO(histogram[histogram.length - 1].month);

  const firstFormatted = format(firstDate, 'MMMM y');
  const lastFormatted = format(lastDate, 'MMMM y');

  if (isEqual(firstDate, lastDate)) {
    return `in ${firstFormatted}`;
  } else if (isSameYear(firstDate, lastDate)) {
    return `between ${format(firstDate, 'MMMM')} and ${lastFormatted}`;
  } else {
    return `between ${firstFormatted} and ${lastFormatted}`;
  }
};

const historyText = ({ passenger }) => {
  return `Appears in the Epstein flight logs ${passenger.flightCount} times ${range(
    passenger.histogram,
  )}`;
};

export default ({ match }) => {
  const { slug } = match.params;
  const { loading, error, data, fetchMore } = useQuery(PASSENGER, {
    variables: { slug, limit: FLIGHT_LIMIT },
  });

  if (loading) return <Loading text />;
  if (error) return <p>{error.message}</p>;

  return (
    <>
      <MetaTags
        title={data.passenger.name}
        uri={`passenger/${data.passenger.slug}`}
        description={`${data.passenger.name} ${historyText(data).replace(
          /(^[A-Z])/,
          cap => cap.toLowerCase(),
        )}`}
      />
      <Details>
        <span>name</span>
        <span>{data.passenger.name}</span>
        <span>bio</span>
        <div>
          <span>{data.passenger.biography}</span>
          <br />
          <a className={link} href={data.passenger.wikipedia_link}>
            <span>See more on Wikipedia â†’</span>
          </a>
        </div>
        <span>history</span>
        <div>{historyText(data)}</div>
        {data.passenger.image ? (
          <>
            <span>image</span>
            <img src={data.passenger.image} alt="" />
          </>
        ) : null}
        <span>flights</span>
        <PaginatedBrowser
          browserComponent={MiniFlight}
          ids={data.passenger.flights.edges.map(x => x.id)}
          totalAvailable={data.passenger.flightCount}
          pageSize={FLIGHT_LIMIT}
          fetchMore={() =>
            fetchMore({
              variables: { offset: data.passenger.flights.edges.length + 1 },
              updateQuery: (prev, { fetchMoreResult }) => {
                if (!fetchMoreResult) return prev;
                return {
                  passenger: {
                    ...prev.passenger,
                    flights: {
                      ...fetchMoreResult.passenger.flights,
                      pageInfo: {
                        ...fetchMoreResult.passenger.flights.pageInfo,
                        count:
                          prev.passenger.flights.pageInfo.count +
                          fetchMoreResult.passenger.flights.pageInfo.count,
                      },
                      edges: [
                        ...prev.passenger.flights.edges,
                        ...fetchMoreResult.passenger.flights.edges,
                      ],
                    },
                  },
                };
              },
            })
          } // eslint-disable-line
        />
      </Details>
    </>
  );
};
