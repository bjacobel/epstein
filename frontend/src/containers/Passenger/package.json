{
  "main": "./Passenger.js",
  "private": true
}
