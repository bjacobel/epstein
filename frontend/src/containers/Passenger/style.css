.skinnyResults {
  max-width: 450px;
}
