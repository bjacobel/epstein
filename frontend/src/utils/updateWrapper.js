import { act } from 'react-dom/test-utils';

export function wait(amount = 0) {
  return new Promise(resolve => setTimeout(resolve, amount));
}

export default async (wrapper, amount = 0) => {
  await act(async () => {
    jest.useRealTimers();
    await wait(amount);
    jest.useFakeTimers();
    wrapper.update();
  });
};
