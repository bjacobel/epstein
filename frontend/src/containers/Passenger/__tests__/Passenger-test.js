import React from 'react';
import { ApolloProvider, ApolloClient, InMemoryCache } from '@apollo/client';
import { mount } from 'enzyme';

import Passenger, { PASSENGER, range } from '../Passenger';
import MockLink from '../../../utils/testing/MockLink';
import updateWrapper from '../../../utils/testing/updateWrapper';

jest.mock('../../../components/MetaTags', () => () => null);

const passengerData = {
  data: {
    passenger: {
      id: 1,
      slug: 'bill-clinton',
      name: 'Bill Clinton',
      biography: 'prez',
      wikipedia_link: 'wikipedia dot com',
      image: 'jpeg',
      flightCount: 26,
      histogram: [
        { month: '2002-02-01', count: 2 },
        { month: '2002-03-01', count: 4 },
      ],
      flights: { edges: [], pageInfo: { count: 10, hasNext: true } },
    },
  },
};

describe('passenger details container', () => {
  let client, link;

  beforeEach(() => {
    link = new MockLink();
    client = new ApolloClient({
      cache: new InMemoryCache({
        addTypename: false,
      }),
      link,
    });
  });

  it('queries by slug from params', async () => {
    const queryHandler = jest.fn(() => new Promise(() => {}));
    link.setRequestHandler(PASSENGER, queryHandler);
    const slug = 'bill-clinton';
    const wrapper = mount(
      <ApolloProvider client={client}>
        <Passenger match={{ params: { slug } }} />
      </ApolloProvider>,
    );
    await updateWrapper(wrapper);

    expect(queryHandler).lastCalledWith(expect.objectContaining({ slug }));
  });

  it('renders spinner before gql loads', async () => {
    const queryHandler = jest.fn(() => new Promise(() => {}));
    link.setRequestHandler(PASSENGER, queryHandler);

    const wrapper = mount(
      <ApolloProvider client={client}>
        <Passenger match={{ params: { slug: 'bill-clinton' } }} />
      </ApolloProvider>,
    );

    await updateWrapper(wrapper);

    expect(wrapper.find(Passenger)).toMatchSnapshot();
  });

  it('matches snapshot with data resolution', async () => {
    const queryHandler = jest.fn().mockResolvedValue(passengerData);
    link.setRequestHandler(PASSENGER, queryHandler);
    const wrapper = mount(
      <ApolloProvider client={client}>
        <Passenger match={{ params: { slug: 'bill-clinton' } }} />
      </ApolloProvider>,
    );

    await updateWrapper(wrapper);

    expect(wrapper.find(Passenger)).toMatchSnapshot();
  });

  describe('range', () => {
    it('omits the first year if the range is all the same year', () => {
      expect(
        range([
          { month: '2002-02-01', count: 1 },
          { month: '2002-04-01', count: 1 },
        ]),
      ).toMatchInlineSnapshot(`"between February and April 2002"`);
    });

    it('omits the second date if there was only one flight month', () => {
      expect(range([{ month: '2002-02-01', count: 1 }])).toMatchInlineSnapshot(
        `"in February 2002"`,
      );
    });

    it('formats both dates if there is a real range', () => {
      expect(
        range([
          { month: '2002-02-01', count: 1 },
          { month: '2002-04-01', count: 1 },
          { month: '2002-08-01', count: 1 },
          { month: '2003-01-01', count: 1 },
          { month: '2003-02-01', count: 1 },
        ]),
      ).toMatchInlineSnapshot(`"between February 2002 and February 2003"`);
    });
  });
});
