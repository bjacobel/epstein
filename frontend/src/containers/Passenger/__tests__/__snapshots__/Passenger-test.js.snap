// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passenger details container matches snapshot with data resolution 1`] = `
<_default
  match={
    Object {
      "params": Object {
        "slug": "bill-clinton",
      },
    }
  }
>
  <mockConstructor
    description="Bill Clinton appears in the Epstein flight logs 26 times between February and March 2002"
    title="Bill Clinton"
    uri="passenger/bill-clinton"
  />
  <_default>
    <div
      className="container"
    >
      <span>
        name
      </span>
      <span>
        Bill Clinton
      </span>
      <span>
        bio
      </span>
      <div>
        <span>
          prez
        </span>
        <br />
        <a
          className="link"
          href="wikipedia dot com"
        >
          <span>
            See more on Wikipedia →
          </span>
        </a>
      </div>
      <span>
        history
      </span>
      <div>
        Appears in the Epstein flight logs 26 times between February and March 2002
      </div>
      <span>
        image
      </span>
      <img
        alt=""
        src="jpeg"
      />
      <span>
        flights
      </span>
      <div
        className="skinnyResults"
      >
        <_default
          browserComponent={[Function]}
          fetchMore={[Function]}
          ids={Array []}
          pageSize={10}
          totalAvailable={26}
        >
          <div
            className="browser"
          >
            <div
              className="controls"
            >
              <button
                aria-controls="detailsList"
                aria-label="previous page of flights"
                className="pageControl"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                ◀ prev
              </button>
              <span
                className="pageCount"
              >
                1 / 3
              </span>
              <button
                aria-controls="detailsList"
                aria-label="next page of flights"
                className="pageControl"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                next ▶
              </button>
            </div>
            <div
              className="hidden"
            >
              <_default>
                <div
                  className="spinWrap"
                >
                  <div
                    className="spinner"
                  />
                </div>
              </_default>
            </div>
            <div
              aria-live="polite"
              className=""
              id="detailsList"
              role="region"
            />
          </div>
        </_default>
      </div>
    </div>
  </_default>
</_default>
`;

exports[`passenger details container renders spinner before gql loads 1`] = `
<_default
  match={
    Object {
      "params": Object {
        "slug": "bill-clinton",
      },
    }
  }
>
  <_default
    text={true}
  >
    <div
      className="spinWrap"
    >
      <div
        className="spinner"
      />
      <span
        className="spinText"
      >
        Hang tight!
      </span>
    </div>
  </_default>
</_default>
`;
